File: scripts/nups_cli_rust/src/ups.rs
use byteorder::ReadBytesExt;
use std::io::Cursor;

use crate::byte_conversion;
use crate::crc32::Crc32;

#[derive(Debug)]
pub struct Ups {
  valid_patch_: bool,
  old_file_size_: u32,
  new_file_size_: u32,
  changed_offset_list_: Vec<u32>,
  xor_bytes_list_: Vec<Vec<u8>>,
  current_ptr_: u32,
  crc32_struct_: Crc32Nums,
  crc32_: Crc32,
}

#[derive(Debug)]
struct Crc32Nums {
  original_file: u32,
  new_file: u32,
  patch: u32,
}

impl Ups {
  pub fn new(crc32: Crc32) -> Ups {
    Ups {
      valid_patch_: false,
      old_file_size_: 0,
      new_file_size_: 0,
      changed_offset_list_: Vec::<u32>::new(),
      xor_bytes_list_: Vec::<Vec<u8>>::new(),
      current_ptr_: 0,
      crc32_struct_: Crc32Nums {
        original_file: 0,
        new_file: 0,
        patch: 0,
      },
      crc32_: crc32,
    }
  }

  pub fn is_valid_patch(&mut self, ups_file: &[u8]) -> Result<(), anyhow::Error> {
    println!("Checking if patch is valid.");
    self.changed_offset_list_.clear();
    self.xor_bytes_list_.clear();

    let mut header = String::from("");

    while self.current_ptr_ < 4 {
      header.push(ups_file[self.current_ptr_ as usize] as char);
      self.current_ptr_ += 1;
    }

    if header != "UPS1" {
      return Err(anyhow::anyhow!("ERROR: Header is not 'UPS1'"));
    }

    self.old_file_size_ = self.decrypt(ups_file);
    self.new_file_size_ = self.decrypt(ups_file);

    let mut file_position: u32 = 0;
    let end_of_file_crc_bytes: u32 = 12;
    let end_ptr = ups_file.len() as u32 - end_of_file_crc_bytes;

    self.set_changed_offsets_and_xor_bytes(ups_file, &mut file_position, &end_ptr);
    self.check_valid_base_on_crc32(ups_file)
  }

  pub fn is_file_valid_to_apply(&self, gba_file: &[u8]) -> Result<(), anyhow::Error> {
    println!("Checking if GBA file is valid...");
    let file_crc32 = self.crc32_.crc32_calculate(gba_file);
    let gba_filelength = gba_file.len() as u32;
    let fit_as_old =
      self.old_file_size_ == gba_filelength && file_crc32 == self.crc32_struct_.original_file;
    let fit_as_new =
      self.new_file_size_ == gba_filelength && file_crc32 == self.crc32_struct_.new_file;

    let results = self.valid_patch_ && (fit_as_old || fit_as_new);
    println!("FINISHED: Checking if GBA file is valid...");
    match results {
      true => Ok(()),
      false => Err(anyhow::anyhow!("GBA file is NOT valid!")),
    }
  }

  pub fn apply_patch(&self, gba_file: &[u8]) -> Vec<u8> {
    let length = std::cmp::max(gba_file.len(), self.new_file_size_ as usize);
    let mut patched_gba_file = vec![0; length];
    let min_length = std::cmp::min(gba_file.len(), length);

    for i in 0..min_length {
      patched_gba_file[i] = gba_file[i];
    }

    for i in 0..self.changed_offset_list_.len() {
      for j in 0..self.xor_bytes_list_[i].len() {
        let index = self.changed_offset_list_[i] as usize + j;
        patched_gba_file[index] ^= self.xor_bytes_list_[i][j];
      }
    }

    patched_gba_file
  }

  // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-
  // | PRIVATE FUNCTIONS
  // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

  fn decrypt(&mut self, ups_file: &[u8]) -> u32 {
    let mut value: u32 = 0;
    let mut shift: u32 = 1;
    let mut x = ups_file[self.current_ptr_ as usize];
    self.current_ptr_ += 1;
    value += (x & 0x7F) as u32 * shift;

    while (x & 0x80) == 0 {
      shift <<= 7;
      value += shift as u32;
      x = ups_file[self.current_ptr_ as usize];
      self.current_ptr_ += 1;
      value += (x & 0x7F) as u32 * shift;
    }

    value
  }

  fn encrypt(&self, offset: u32) -> Vec<u8> {
    let mut bytes = Vec::<u8>::new();
    let mut x: u32 = offset & 0x7F;
    let mut offset: u32 = offset >> 7;

    while offset != 0 {
      bytes.push(x as u8);
      offset -= 1;
      x = offset & 0x7F;
      offset >>= 7;
    }

    let final_byte = (0x80 | x) as u8;
    bytes.push(final_byte);
    bytes
  }

  fn set_ups_crc32(&mut self, ups_file: &[u8]) {
    let ptr = self.current_ptr_ as usize;

    let mut cursor = Cursor::new(&ups_file[ptr + 0..ptr + 4]);
    self.crc32_struct_.original_file = cursor.read_u32::<byteorder::LittleEndian>().unwrap();

    let mut cursor = Cursor::new(&ups_file[ptr + 4..ptr + 8]);
    self.crc32_struct_.new_file = cursor.read_u32::<byteorder::LittleEndian>().unwrap();

    let mut cursor = Cursor::new(&ups_file[ptr + 8..ptr + 12]);
    self.crc32_struct_.patch = cursor.read_u32::<byteorder::LittleEndian>().unwrap();
  }

  fn to_binary(&self) -> Vec<u8> {
    let mut byte_vector = Vec::<u8>::new();
    byte_vector.push('U' as u8);
    byte_vector.push('P' as u8);
    byte_vector.push('S' as u8);
    byte_vector.push('1' as u8);

    byte_vector.extend(&self.encrypt(self.old_file_size_));
    byte_vector.extend(&self.encrypt(self.new_file_size_));

    for i in 0..self.changed_offset_list_.len() {
      let mut relative_offset = self.changed_offset_list_[i];
      if i != 0 {
        let temp_offset =
          self.changed_offset_list_[i - 1] + (self.xor_bytes_list_[i - 1].len() as u32) + 1;
        relative_offset -= temp_offset;
      }

      byte_vector.extend(&self.encrypt(relative_offset));
      byte_vector.extend(&self.xor_bytes_list_[i]);
      byte_vector.push(0u8);
    }

    byte_vector.extend(byte_conversion::convert_int_to_bytes_little_endian(
      self.crc32_struct_.original_file,
    ));

    byte_vector.extend(byte_conversion::convert_int_to_bytes_little_endian(
      self.crc32_struct_.new_file,
    ));

    byte_vector
  }

  fn set_changed_offsets_and_xor_bytes(
    &mut self,
    ups_file: &[u8],
    file_position: &mut u32,
    end_ptr: &u32,
  ) {
    while self.current_ptr_ + 1 < *end_ptr {
      *file_position += self.decrypt(ups_file);
      self.changed_offset_list_.push(*file_position as u32);
      let mut new_xor_data = Vec::<u8>::new();

      while ups_file[self.current_ptr_ as usize] != 0 {
        new_xor_data.push(ups_file[self.current_ptr_ as usize]);
        self.current_ptr_ += 1;
      }

      let new_xor_len = new_xor_data.len();
      self.xor_bytes_list_.push(new_xor_data);
      *file_position += (new_xor_len as u32) + 1;
      self.current_ptr_ += 1;
    }
  }

  fn check_valid_base_on_crc32(&mut self, ups_file: &[u8]) -> Result<(), anyhow::Error> {
    self.set_ups_crc32(ups_file);
    let binary_byte_vector = self.to_binary();
    let calculated_crc32 = self.crc32_.crc32_calculate(&binary_byte_vector);
    self.valid_patch_ = calculated_crc32 == self.crc32_struct_.patch;
    println!("FINISHED: Checking if patch is valid.");
    match self.valid_patch_ {
      true => Ok(()),
      false => Err(anyhow::anyhow!("Patch was not valid")),
    }
  }
}

File: scripts/nups-cli-cpp/src/ups.cpp
#include "ups.hpp"

Ups::Ups() :
  validPatch_(false),
  old_file_size_(0), new_file_size_(0) {

}

bool Ups::is_valid_patch(std::vector<uint8_t> &ups_file) {
  this->changed_offset_list_.clear();
  this->xor_bytes_list_.clear();

  uint8_t *ups_ptr = &(ups_file[0]);
  uint8_t *current_ptr = ups_ptr;

  std::string header;
  for (int i = 0; i < 4; i++) {
    header.push_back(*current_ptr);
    current_ptr += 1;
  }

  if (header.compare("UPS1") != 0)
    return false;

  this->old_file_size_ = this->decrypt(&current_ptr);
  this->new_file_size_ = this->decrypt(&current_ptr);

  // Body, refactor here! TODO
  unsigned long file_position = 0;
  unsigned int end_of_file_crc_bytes = 12;
  unsigned int end_ptr = static_cast<unsigned int>(ups_file.size()) - end_of_file_crc_bytes;

  while (current_ptr - ups_ptr + 1 < end_ptr) {
    file_position += this->decrypt(&current_ptr);
    this->changed_offset_list_.push_back(file_position);
    std::vector<uint8_t> new_xor_data;

    while (*current_ptr != 0) {
      new_xor_data.push_back(*current_ptr);
      current_ptr += 1;
    }
    this->xor_bytes_list_.push_back(new_xor_data);
    file_position += static_cast<unsigned long>(new_xor_data.size()) + 1;
    current_ptr += 1;
  }

  this->set_ups_crc32(current_ptr, ups_file);

  auto binary_byte_vector = this->to_binary();
  unsigned int calculated_crc32 = this->crc32_.crc32_calculate(
                                    binary_byte_vector);

  this->validPatch_ = this->patch_crc32_ == calculated_crc32;

  return this->validPatch_;
}

bool Ups::is_file_valid_to_apply(const std::vector<uint8_t> &gba_file) {
  std::cout << "Checking if GBA file is valid" << std::endl;

  unsigned int file_crc32 = this->crc32_.crc32_calculate(gba_file);
  unsigned long gba_filelength = static_cast<unsigned long>(gba_file.size());
  bool fit_as_old = (this->old_file_size_ == gba_filelength)
                    && (file_crc32 == this->original_file_crc32_);
  bool fit_as_new = (this->new_file_size_ == gba_filelength)
                    && (file_crc32 == this->new_file_crc32_);

  std::cout << "Finished: Checking if GBA file is valid" << std::endl;
  return this->validPatch_ && (fit_as_old || fit_as_new);
}

std::vector<uint8_t> Ups::apply_patch(const std::vector<uint8_t> &gba_file) {
  unsigned long length = std::max<unsigned long>(static_cast<unsigned long>
                         (gba_file.size()), this->new_file_size_);
  std::vector<uint8_t> result(length);
  unsigned long min_length = std::min<unsigned long>(static_cast<unsigned long>
                             (gba_file.size()), static_cast<unsigned long>(result.size()));

  uint8_t *result_ptr = &result[0];

  for (unsigned int i = 0; i < min_length; i++)
    result_ptr[i] = gba_file[i];

  for (size_t i = 0; i < this->changed_offset_list_.size(); i++) {
    for (size_t xor_bytes_len = 0; xor_bytes_len < this->xor_bytes_list_[i].size();
         xor_bytes_len++) {
      auto ulong_len = static_cast<unsigned long>(xor_bytes_len);
      result_ptr[this->changed_offset_list_[i] + ulong_len] ^=
        this->xor_bytes_list_[i][xor_bytes_len];
    }
  }

  return result;
}

// =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
// | PRIVATE FUNCTIONS
// =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

unsigned long Ups::decrypt(uint8_t *pointer[]) {
  unsigned long value = 0;
  int shift = 1;
  uint8_t x = **pointer;
  *pointer += 1;
  value += static_cast<unsigned long>((x & 0x7F) * shift);

  while ((x & 0x80) == 0) {
    shift <<= 7;
    value += static_cast<unsigned long>(shift);
    x = **pointer;
    *pointer += 1;
    value += static_cast<unsigned long>((x & 0x7F) * shift);
  }

  return value;
}

std::vector<uint8_t> Ups::encrypt(unsigned long offset) {
  std::vector<uint8_t> bytes;

  unsigned long x = offset & 0x7F;
  offset >>= 7;
  while (offset != 0) {
    bytes.push_back(static_cast<uint8_t>(x));
    offset -= 1;
    x = offset & 0x7F;
    offset >>= 7;
  }

  bytes.push_back(static_cast<uint8_t>(0x80 | x));
  return bytes;
}

std::vector<uint8_t> Ups::to_binary() {
  std::vector<uint8_t> byte_vector;
  byte_vector.push_back('U');
  byte_vector.push_back('P');
  byte_vector.push_back('S');
  byte_vector.push_back('1');

  vector_concat::concat(byte_vector, this->encrypt(this->old_file_size_));
  vector_concat::concat(byte_vector, this->encrypt(this->new_file_size_));

  for (int i = 0; i < this->changed_offset_list_.size(); i++) {
    unsigned long relative_offset = this->changed_offset_list_[i];
    if (i != 0) {
      auto temp_offset = this->changed_offset_list_[i - 1]
                         + this->xor_bytes_list_[i - 1].size() + 1;
      relative_offset -= static_cast<unsigned long>(temp_offset);
    }

    vector_concat::concat(byte_vector, this->encrypt(relative_offset));
    vector_concat::concat(byte_vector, this->xor_bytes_list_[i]);
    byte_vector.push_back(0);
  }

  std::vector<uint8_t> temp_byte_array(4);
  temp_byte_array = byte_conversion::int_to_bytes_little_endian(this->original_file_crc32_);
  vector_concat::concat(byte_vector, temp_byte_array);
  temp_byte_array = byte_conversion::int_to_bytes_little_endian(this->new_file_crc32_);
  vector_concat::concat(byte_vector, temp_byte_array);

  return byte_vector;
}

void Ups::set_ups_crc32(const uint8_t* current_ptr,
                        const std::vector<uint8_t> &ups_file) {
  const uint8_t* ups_ptr = &ups_file[0];
  unsigned int start_index = static_cast<unsigned int>(current_ptr - ups_ptr);
  unsigned int byte_length = 4;
  byte_conversion::bytes_to_uint_little_endian(
    ups_file, start_index + 8, 4);
  this->original_file_crc32_ = byte_conversion::bytes_to_uint_little_endian(
                                 ups_file, start_index + 0, 4);
  this->new_file_crc32_ = byte_conversion::bytes_to_uint_little_endian(
                            ups_file, start_index + 4, 4);
  this->patch_crc32_ = byte_conversion::bytes_to_uint_little_endian(
                         ups_file, start_index + 8, 4);
}
